driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+class BatchCursorFlux<T> implements Publisher<T> {
+
+    private final BatchCursorPublisher<T> batchCursorPublisher;
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+    private final AtomicLong demandDelta = new AtomicLong(0);
+    private BatchCursor<T> batchCursor;
+    private FluxSink<T> sink;
+
+    BatchCursorFlux(final BatchCursorPublisher<T> batchCursorPublisher) {
+        this.batchCursorPublisher = batchCursorPublisher;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        Flux.<T>create(sink -> {
+            this.sink = sink;
+            sink.onRequest(demand -> {
+                if (calculateDemand(demand) > 0 && inProgress.compareAndSet(false, true)) {
+                    if (batchCursor == null) {
+                        int batchSize = calculateBatchSize(sink.requestedFromDownstream());
+                        batchCursorPublisher.batchCursor(batchSize).subscribe(bc -> {
+                            batchCursor = bc;
+                            inProgress.set(false);
+                            recurseCursor();
+                        }, sink::error);
+                    } else {
+                        inProgress.set(false);
+                        recurseCursor();
+                    }
+                }
+            });
+            sink.onCancel(this::closeCursor);
+            sink.onDispose(this::closeCursor);
+        }, FluxSink.OverflowStrategy.BUFFER)
+        .subscribe(subscriber);
+    }
+
+    private void closeCursor() {
+        if (batchCursor != null && !batchCursor.isClosed()) {
```

Comment:
Do we want to document `BatchCursor.close` as idempotent and spare ourselves from checks like this one? Its implementation is both not idempotent (not required, but could be) and incorrect at this moment, but the bug will be fixed once https://github.com/mongodb/mongo-java-driver/pull/654 is merged, and the method will also become idempotent.

Reply:
Updated